### PR TITLE
Tutorial for External Controller Python Template

### PR DESCRIPTION
**Pull request by andermi**
https://osrf.github.io/buoy_entrypoint/pr-preview/pr-25/Tutorials/ROS2/PythonTemplate/

part of https://github.com/osrf/buoy_entrypoint/issues/27

### CODE REVIEW

File: docs/docs/tutorials.md
Comment (mabelzhang) [inline docs/docs/tutorials.md:17]: Do these two files exist yet? They're not added to this PR. Should commit at the same time as the files.

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:6]:
```suggestion
set up (C++/Python) for a quick start on writing a
```

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:39]: This doesn't look indented in the preview
```suggestion
   `$ git clone https://github.com/<owner>/<repo_name>.git`
```

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:86]: These line numbers could change if the files are changed in the template, right? Then the tutorial will have to be updated as well, not just one number, but all numbers when any line is added? Is there a way to make this less specific and more future-proof, so that people don't have to keep remembering to update in both places?

The other consideration is that, since the code block below doesn't display line numbers, it's hard for people to tell what these line numbers are talking about. It could be more clear to instead have in-line comments in the code saying, change this line, it does X, so that people understand what the code is doing.

Same for all code blocks on this page with line number references.

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:4]: Some high-level intro of what this tutorial will enable people to do would be helpful. It reads very technical, which if someone doesn't understand the purpose, it just dives straight in.

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:323]: Is this page meant to be added in this PR? If not, this link will appear like a link that goes nowhere. It should be added when that page exists.

**Review by mabelzhang — COMMENTED** [review]
Nit: Github -> GitHub, python -> Python, ROS2 -> ROS 2 (the correct stylistic name has a space).

I did a conceptual read through. I have not actually followed the tutorial yet. We should definitely have at least one person who didn't write the tutorial try to follow it and see if they can make it work. That could be me, unless someone gets to it first.

**Comment by mabelzhang** [comment]
Oh yeah, while we're here, could you add a fix for the two `conveter` typos in the first sentence of the first page in the documentation :sweat_smile: Not a good first impression.
https://osrf.github.io/buoy_entrypoint/

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (andermi) [reply docs/docs/Tutorials/ROS2/PythonTemplate.md:86]: It shows line numbers for me... are there really no line numbers showing up?

I tried highlighting lines but for some reason the `hl_lines` instruction in the code block isn't actually highlighting... so I just put line numbers. Also, since markdown doesn't allow referencing an actual file like .rst, this tutorial would need to be updated with any code changes anyways... But maybe the compromise is to have inline comments like you suggest

Comment (andermi) [reply docs/docs/Tutorials/ROS2/PythonTemplate.md:92]: added inline comments in code blocks. Need to confirm that others can see the line numbers in the code blocks

**Comment by andermi** [comment]
@mabelzhang @quarkytale @hamilton8415 Would anyone like to try to run through the tutorial to make sure it works? And then we can get it merged in

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:44]: I think getting rid of the newline will make the code snippet display correctly under the bullet
```suggestion
```

File: README.md
Comment (mabelzhang) [inline README.md:31]: 404 on this link. Do you mean this?
```suggestion
[python examples](https://github.com/osrf/buoy_msgs/tree/main/buoy_api_py/buoy_api/examples) for example
```

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:17]: 404
```suggestion
[Python examples](https://github.com/osrf/buoy_msgs/tree/main/buoy_api_py/buoy_api/examples) for example
```

File: docs/docs/Tutorials/ROS2/PythonTemplate.md
Comment (mabelzhang) [inline docs/docs/Tutorials/ROS2/PythonTemplate.md:273]: Do you mean ControlPolicy.target? policy is just a variable
```suggestion
The `ControlPolicy.target` function may be called from within the `Controller` class. You may call it inside any of
```

**Review by mabelzhang — APPROVED** [review]
Minor fixes, then good to merge.

I've read through the tutorial and created a repo from the template as instructed. Seems straight-forward to me, but I'm not a controls researcher, so maybe I don't fit the target audience :) I'm just the QA.

It'd be nice to include a link to this tutorial from the Python template repo, once this tutorial is merged.

**Pull request merged**
